dp/generate_climos.py

```diff
@@ -389,7 +440,8 @@ def update_metadata_and_time_var(input_file, t_start, t_end, operation, climo_fi
         # Update frequency attribute to reflect that this is a climo file.
         suffix = {
             'std': 'SD',
-            'mean': 'Mean'
+            'mean': 'Mean',
+            'sum': 'Mean'
```

Comment:
Same as above, I think we have to consider the multiple operations.  Mostly because `frequency` is used later on in `nchelpers` to create a `unique_id`.  If we have a degree day file that uses `-p sum` and another that uses `-p mean` they will produce identical `unique_ids` even though they are not the same.  Thus, the indexing script will just replace one file with the other, rather than keep both (I just had this issue with gridded climatologies :cry:).

Looking at the [PCIC metadata standard](https://pcic.uvic.ca/confluence/pages/viewpage.action?spaceKey=CSG&title=PCIC+metadata+standard+for+downscaled+data+and+hydrology+modelling+data) (in the **PCIC standard frequency attribute values** section), the `frequency` attribute has an `<operator>` associated with it.  So potentially that description could be extended to include multiple operations (i.e. `mClimSumMean`) but I think that is a slippery slope.  My gut says that the suffix should be `sum` however, that too could cause problems in the future (i.e.  if we **just** wanted to sum the data).

I think this requires some discussion, perhaps @rod-glover could help us here.

Reply:
Hm. I guess I'm starting to feel like I took the wrong approach to this problem. This isn't really a "sum" operation, despite me making it attached to the "operation" flag, more like a mean that CDO isn't smart enough to calculate in a single step, that I borrowed the nice stdev operation infrastructure to add. Adequately effective, but not the right approach.

Maybe, like @rod-glover suggested in our meeting yesterday, it should be a `data-type` flag - data either represents a (possibly simulated) **measurement**, in which case yearly values are the means of daily values, or a **count** of the number of times an event occurs, in which yearly values are the sums of daily values. Measurement would be the default data type.

That gives me an extra headache, though: I would have to figure out how to make cdo calculate the standard deviation of  "count" data types. It would be a lot of extra code, but maybe code that better reflected the mathematical reality.

Reply:
Well, here's a nice philosophical question: In which of the following ways do we regard the final product? Because that is what both `operation` and `frequency` signify.

1. It is the climatological mean of an abstract variable that spans years and that _just happens, as an insignificant computational detail_ to be computed from finer-resolution inputs along the way.
1. It is a more complex statistic that must fundamentally be regarded as the (mean | sum) within years followed by mean over years on the finer-resolution variables.

Let's consider the arguments for both cases in the context of count variables. 

1. The abstract variable is a count variable, which is necessarily counted (summed) over a specified period (e.g., day, month, season, year). In that case, "sum within _period_" is just part of the definition of the variable of interest, and climatology (mean over years) of the variable of interest is the operation we are actually interested in.
1. The abstract variable is still a count variable, but it is counted (summed) over a shorter period and must be further processed to reflect the real period of interest. The operation we are interested in is actually this process of both summing up to the period of interest and then forming climatologies.

(The arguments are the same for point/continuous variables, but the nuances get obscured by the commonality of the operation mean in both the within-period and across-years operations.)

I think case 2 is significantly weaker. In which case, this is just a (climatological) mean, and the grotty details of computation are not to be placed in the filename/unique id. And (phew) we do not need to consider using compound methods in the `frequency` attribute just yet. `xClimMean` will continue to suffice.

However, @nikola-rados , your observation about `-p sum` and `-p mean` operations yielding the same unique ids is important. It's important because `-p mean` (in the sense of mean within years) _is invalid_ for degree-day and other count variables. And therefore we need, at minimum, to add operation checking that ensures that invalid combinations of variables and operations aren't allowed to proceed. 

Even better, I think we should eliminate the "sum" operation, and always call it "mean" (signifying _climatological_ mean, which is actually the point of the computation). The _code_, not the user, decides how to compute that climatology, based, in part, on the kind of variable it is processing. In the case of point variables (e.g., temperature), it is mean of means; in the case of counts (e.g., gdd), it is mean of sums. This would eliminate the possibility of user-induced error. 

This will simplify some parts of the code while introducing some complications elsewhere. I believe those complications are quite justified, and less inclined to lead us into problems later.

To sum up, we need to distinguish between _climatological_ operation (mean or standard deviation) and intermediate computational operation (sum or mean, dependent on variable type) that achieves the desired end given our inputs. It so happens that one case is prepackaged for us by CDO, and the other must be specially coded (by Lee). This is a very desirable distinction to make, because one is specification (climatological operation), and the other is means/implementation. In general we want to highlight the former and hide the latter.

Reply:
Aha, @corviday , I think you arrived at much the same conclusion as I did.

Reply:
And you are right: Standard deviation of count data types may take some complicated code. But it's very worthwhile to consider that case, because it foregrounds the distinctions that got lost in case of mean.

Questions:

1. What actually are we taking the standard deviation of in the case of point/continuous variables (e.g., temperature)? Is it:
   1. SD(means of observations within years), taken across years
   1. SD(all observations)

1. What are we taking standard deviations of in the case of count variables (e.g., gdd)? Is it:
   1. SD(sums of observations within years), taken across years
   1. SD(all observations)

For both types of variables, the two different SD's are mathematically different, i.e., give different values. This is unlike the case for means, where the formulas are essentially the same.